zkvm/docs/zkvm-blockchain.md

```diff
@@ -0,0 +1,406 @@
+# ZkVM blockchain specification
+
+This is the specification for the ZkVM blockchain,
+a blockchain containing
+[ZkVM transactions](zkvm-spec.md).
+
+Nodes participating in a ZkVM blockchain network must implement the data types and perform the procedures described in this document.
+Specifically:
+
+- Each node maintains a [blockchain state](#blockchain-state).
+- A node creating a new ZkVM network performs the [start new network](#start-new-network) procedure.
+- A node joining an existing ZkVM network performance the [join existing network](#join-existing-network) procedure.
+- Each node performs the [apply block](#apply-block) procedure on the arrival of each new block.
+
+This document does not describe the consensus mechanism,
+which governs the production of authoritative blocks for the blockchain.
+
+_TBD: add a consensus spec._
+
+# Data types
+
+## Blockchain state
+
+The state of a ZkVM blockchain is given by the blockchain-state structure.
+Each node maintains a copy of this structure.
+As each new
+[block](#block)
+arrives,
+it is
+[applied](#apply-block)
+to the current state to produce an new,
+updated state.
+
+The blockchain state contains:
+
+- `initialheader`:
+  The initial block header
+  (the header with height 1).
```

Comment:
I never liked the kr's argument that number 0 is useful to reserve for "nil" case. In Rust there's Option for exactly that, and we all know that indexing from 1 always strikes back.

Reply:
```suggestion
  (the header with height 0).
```

Reply:
In fact I think it was I who first lobbied for the initial block to have height 1 instead of 0, and ISTR several things falling nicely into place once we made that change. I can think of no situation from then until now in which it has "struck back." I still feel strongly that 1 is the right choice.

Remember that the spec should be language-agnostic.

Reply:
>the spec should be language-agnostic.

I don't think that's a good argument — similar to LE/BE, most computers are LE and if we just make things BE, it's going to induce more confusion for everyone. Similarly, virtually all languages that matter today index things from 0, and things "fall in place" pretty nicely that way. Having height starting with 1 is an oddity that must be justified with a strong reason, not the other way around. If the only reason is that in Go there is no Option<T> type, while everywhere else there is (NULL in SQL DBs, JS/Ruby having null/nil, optionals in Rust/C++/Swift etc), it's a problem specific to Go and must be worked around there.

Can you remind what was improved by having the height start at 1?

Reply:
Bitcoin uses height=0: https://blockstream.info/block/000000000019d6689c085ae165831e934ff763ae46a2a6c172b3f1b60a8ce26f
(as well as its codebase forks like Zcash, Monero etc)

Ethereum uses height=0: https://etherscan.io/block/0xd4e56740f876aef8c010b86a40d5f56745a118d0906a34e69aec8c0db1cb8fa3

Stellar uses height=1: https://www.stellar.org/developers/guides/concepts/ledger.html

Tezos uses height=0: https://tzscan.io/0

Cosmos uses height=1 for first proposed block, because 0th is a "genesis file": https://www.mintscan.io/blocks/1

Reply:
As someone who knows Cosmos well told me in private conversation: 

> me: do you know why cosmos initial block has height=1, not 0?
> they: Zeroth block is the genesis file
> me: genesis file is somewhat special, or it's just a "baked in block"?
> they: it's special. Just json but I believe you can query the node for height zero state and get genesis info

Reply:
I suggest we use height=0. Another way to look at this: this is the only block that's implicitly agreed upon by everyone, while all the other blocks (with actual transactions) are subject to consensus protocol.

Reply:
i don't care - only delta matters in the end of the day.